rts/motoko-rts/src/gc/mark_compact.rs

```diff
@@ -0,0 +1,312 @@
+            }
+        }
+
+        TAG_ARRAY => {
+            let array = obj as *mut Array;
+            let array_payload = array.payload_addr();
+            for i in 0..array.len() {
+                thread(array_payload.add(i as usize), heap_base);
+            }
+        }
+
+        TAG_MUTBOX => {
+            let mutbox = obj as *mut MutBox;
+            let field_addr = &mut (*mutbox).field;
+            thread(field_addr, heap_base);
+        }
+
+        TAG_CLOSURE => {
+            let closure = obj as *mut Closure;
+            let closure_payload = closure.payload_addr();
+            for i in 0..closure.size() {
+                thread(closure_payload.add(i as usize), heap_base);
+            }
+        }
+
+        TAG_SOME => {
+            let some = obj as *mut Some;
+            let field_addr = &mut (*some).field;
+            thread(field_addr, heap_base);
+        }
+
+        TAG_VARIANT => {
+            let variant = obj as *mut Variant;
+            let field_addr = &mut (*variant).field;
+            thread(field_addr, heap_base);
+        }
+
+        TAG_CONCAT => {
+            let concat = obj as *mut Concat;
+            let field1_addr = &mut (*concat).text1;
+            thread(field1_addr, heap_base);
+            let field2_addr = &mut (*concat).text2;
+            thread(field2_addr, heap_base);
+        }
+
+        TAG_OBJ_IND => {
+            let obj_ind = obj as *mut ObjInd;
+            let field_addr = &mut (*obj_ind).field;
+            thread(field_addr, heap_base);
+        }
+
+        TAG_BITS64 | TAG_BITS32 | TAG_BLOB | TAG_BIGINT => {
+            // These don't have pointers, skip
+        }
+
+        TAG_NULL => {
+            rts_trap_with("encountered NULL object tag in thread_obj_fields");
+        }
+
+        TAG_FWD_PTR | _ => {
+            rts_trap_with("invalid object tag in thread_obj_fields");
+        }
+    }
+}
+
+unsafe fn thread(field: *mut SkewedPtr, heap_base: u32) {
+    if (*field).is_tagged_scalar() {
+        // Field has a scalar value, skip
+        return;
+    }
+
+    if (*field).unskew() < heap_base as usize {
+        // Field doesn't point to heap. The pointee won't be moved so no need to thread the field
+        return;
+    }
+
+    // Store pointed object's header in the field, field address in the pointed object's header
+    let pointed = (*field).unskew() as *mut Obj;
+    let pointed_header = pointed.tag();
+    *field = SkewedPtr(pointed_header as usize);
+    (*pointed).tag = field as u32;
+}
+
+/// Unthread all references, replacing with `new_loc`
+unsafe fn unthread(obj: *mut Obj, new_loc: u32) {
+    // NOTE: For this to work heap addresses need to be greater than the largest value for object
+    // headers. Currently this holds. TODO: Document this better.
+    let mut header = (*obj).tag;
+    while header > TAG_NULL {
+        // TODO: is `header > TAG_NULL` the best way to distinguish a tag from a pointer?
```

Comment:
This is a bit hairy. Maybe leave a note in the cod where we define header tags that these tags must be smaller than any possible dynamic pointer?

Also, this would limit our wiggle room if we want to put variant tags or array lengths etc into the header… (#916)

Reply:
I think the fact that you can distinguish headers from pointers is unfortunately fundamental to the threading algorithm, unless there's some other technique that can be used.

Reply:
Right. It’ll take away roughly ¼ of the possible header values. Still plenty of room for other things.